### ISSUE TITLE
Question about 2dgs u,v center to image_space center in demo notebook

### ISSUE THREAD

**Issue by MazzARQ** [hbb1/2d-gaussian-splatting#182]
Hello! I am playing around with python implementation you provided (thanks for the effort it is super useful) to better understand all the geometry transformations.

`T = M @ projmat # T stands for (WH)^T in Eq.9`

With this, that if I understand correctly is a UV to ImageSpace transformation Matrix, where you removed the column (the row in this case since it's transposed compared to the paper) that is full of zero. so it gets as input u,v,1 and put out xw,yw,w,w.

By doing **0,0,1 @ T** I should get the center of the gaussians in image space right? since all the gaussians are modeled such that 0,0 is their center in the UV plane.

When you are computing the AABB (the part I am trying to understand now), I get different values for the point_image value (following the AABB computation) and center_uv in image space by doing **0,0,1 @ T**.

In my mind I was expecting them to be the same, is there a reason why the two values differs a little bit? Or I am missing something in my understanding?

Thanks for the beautiful work and all the resources

**Comment by hbb1** [comment]
Hi, that's a really good catch! The issue is that the projection of the 3D center is not equal to the center of the projected Axis-Aligned Bounding Box (AABB). 

In most cases, these two centers are quite similar. When the 2D Gaussian is parallel to the image plane, they converge; otherwise, they diverge. 

This discrepancy occurs mainly because, after perspective projection, the **projected 2D Gaussian in screen space** is no longer an ideal **2D Gaussian function** (e.g.,  it cannot be described with 2D covariance). As a result, it becomes asymmetrical, causing the center of the AABB to differ from the 3D center.

**Comment by MazzARQ** [comment]
Thanks for the super quick answer!

So is there a reason why you choose to use the AABB center instead of the Projected one to do all the computations?

**Comment by hbb1** [comment]
The AABB's center is only for low pass filter purpose, which crates a circle on the screen plane. Basically, it covers more effective pixels than that uses the projected 3D center.

**Comment by MazzARQ** [comment]
AAAH now I understand...
In 3DGS you splat and then you check in each pixels how many splats are there. Then you use the splatted center to compute the distance, so in a way their power is computed on the 2D space.

In the 2dgs case you are throwing a ray from 2D to 3D, to find the u,v on the U,V plane, so you compute the power from the 3D/UV space. So the center of the bounding box is not needed for the computation of the power, is just needed to know in which threads/pixel you are going to look for that guassian or not. Even tho it's a little bif off, it doesnt change the correctness of the power calculation, because you don't use it.

I had this confusion because the famous point_image in 3dgs is used to compute the distance, but in 2dgs case it's different

**Comment by hbb1** [comment]
Exactly! We evaluate the 2D Gaussian on object space (for perspective correctness); By contrast, 3DGS slices a 2D Gaussian parallel to the image plane with First order Taylor expansion from its center.

**Comment by MazzARQ** [comment]
Sorry for another question, feels like I am taking advantage of your kindess, as I said before I am trying to project a point from UV to Image space. (something that if I am not mistaken you never do in the cuda code). I was able to have it working on the python version

0,0,1 @ T in the cuda CODE using transMat but I don't get the correct results. I think it should work giving that transMat is UV to Image_Space, but maybe I am missing something.

```
__global__ void computeAxisCUDA(int P,
							glm::mat3 *transMat,
							glm::vec3 *outputs,
							const int *radii)
{
	auto idx = cg::this_grid().thread_rank();
	if (idx >= P)
		return;

	//define 5 glm vector of 3 values
	glm::vec3 center = glm::vec3(0.0, 0.0, 1.0);
	glm::mat3 transmat = transMat[idx];
	
	glm::vec3 transformed_center = center * transmat;
```


I hope in the future I get chance to repay your kindness

**Comment by hbb1** [comment]
I think you need to do the transpose because that **H** is the transpose of **M** where **x=Mu**.

**Comment by hbb1** [comment]
Try this
```c++
glm::vec3 transformed_center = glm::transpose(transmat)  * center; 
// or 
glm::vec3 transformed_center = transmat * center; // forgotten whether it is transposed.
```

**Comment by MazzARQ** [comment]
I was normalizing the homogenous coordinate component after the integer conversion and not before

points_int = points.int()
points_int = points_int[..., :3] / points_int[..., 2:3]

instead of

points = points[..., :3] / points[..., 2:3]
points_int = points.int()

before the plotting, so that's why nothing was working


```
glm::vec3 transformed_center = center * transmat; 
//or
glm::vec3 transformed_center = glm::transpose(transmat)  * center;
```

both works

Thank you very very very much
